File: packages/python/src/sah_sdk/core/bootstrap.py
"""Embedded HMAC bootstrap signature for POST /v1/sessions.

This module pins the byte-exact wire format documented at
``api/guides/authentication-advanced.md#test-vector`` and reproduced in
``tests/conformance/fixtures/auth-vector.json``. The reference TypeScript
implementation lives at ``packages/ts/src/core/internal/auth/bootstrap.ts``;
this file MUST reproduce the identical hex against the same inputs.

Two-stage flow:

1. Serialize the request body as JSON, keys in source order
   (``keycode``, ``email``, ``deviceId``), no whitespace, no trailing
   newline.
2. Build the canonical signing string and HMAC-SHA256 it with the
   ``license_key`` as the key.

Why a dedicated module: the bootstrap signature predates any session (no
``session_secret`` exists yet), uses the ``license_key`` as the HMAC key,
and is the only call where ``device_id`` appears in the body. The
steady-state session-signing helper (``sign_request``) handles every
other call.
"""

from __future__ import annotations

import hashlib
import hmac
import json
from dataclasses import dataclass


@dataclass(frozen=True)
class BootstrapInput:
    """Inputs to the bootstrap signature.

    Mirrors the ``auth-vector`` fixture one-for-one. Field declaration
    order matches the canonical serialized-body order
    (``keycode, email, deviceId``); the body is hand-serialized so the
    Python source order pins the wire bytes.
    """

    keycode: str
    """Per-seat keycode (e.g. ``SDV-HWH-WDD``)."""

    email: str
    """License-owner email (lowercased lookup key server-side)."""

    license_key: str
    """Long-lived license key. HMAC key only — never on the wire."""

    device_id: str
    """Stable per-install device id. Appears in body + ``X-Device-ID``."""

    method: str
    """Uppercase HTTP method, typically ``POST``."""

    path: str
    """Request path with leading ``/v1/``, no query string."""

    timestamp: int
    """Unix seconds. Server tolerates 5 minutes of skew."""


@dataclass(frozen=True)
class BootstrapSignature:
    """Output bundle.

    Returns every intermediate so that conformance tests can assert each
    stage independently — if a future regression flips the
    serialized body, the failure points at exactly that stage instead of
    just "hex differs".
    """

    serialized_body: str
    """JSON body exactly as sent on the wire. Bytes signed verbatim."""

    canonical: str
    """``<ts>.<METHOD> <path>.<body>`` — the HMAC input."""

    hex: str
    """Lowercase hex HMAC-SHA256 over canonical, keyed by license_key."""

    header: str
    """``ISA-Signature: t=<ts>,v1=<hex>`` — ready-to-set header value."""


def build_bootstrap_signature(inp: BootstrapInput) -> BootstrapSignature:
    """Build the byte-exact bootstrap signature.

    Raises:
        ValueError: when any required field is blank or the timestamp is
            non-positive. The locked contract requires every field present.
    """
    if not inp.keycode or inp.keycode != inp.keycode.strip():
        raise ValueError("bootstrap signature: keycode is required")
    if not inp.email or inp.email != inp.email.strip():
        raise ValueError("bootstrap signature: email is required")
    if not inp.license_key or inp.license_key != inp.license_key.strip():
        raise ValueError("bootstrap signature: license_key is required")
    if not inp.device_id or inp.device_id != inp.device_id.strip():
        raise ValueError("bootstrap signature: device_id is required")
    if not inp.method or inp.method != inp.method.strip():
        raise ValueError("bootstrap signature: method is required")
    if not inp.path or inp.path != inp.path.strip():
        raise ValueError("bootstrap signature: path is required")
    if inp.timestamp <= 0:
        raise ValueError("bootstrap signature: timestamp is required")

    serialized_body = _serialize_body(inp.keycode, inp.email, inp.device_id)
    canonical = f"{inp.timestamp}.{inp.method.upper()} {inp.path}.{serialized_body}"
    digest = hmac.new(
        inp.license_key.encode("utf-8"),
        canonical.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()
    return BootstrapSignature(
        serialized_body=serialized_body,
        canonical=canonical,
        hex=digest,
        header=f"ISA-Signature: t={inp.timestamp},v1={digest}",
    )


def _serialize_body(keycode: str, email: str, device_id: str) -> str:
    """Serialize the bootstrap body with pinned key order, no whitespace.

    ``json.dumps`` on a dict preserves insertion order (Python 3.7+); we
    rely on that AND pass ``separators=(',', ':')`` so the JSON is
    minified to match the TypeScript ``JSON.stringify`` output.
    """
    return json.dumps(
        {"keycode": keycode, "email": email, "deviceId": device_id},
        separators=(",", ":"),
        ensure_ascii=False,
    )

File: packages/csharp/src/Core/Bootstrap.cs
// Embedded HMAC bootstrap signature for POST /v1/sessions.
//
// This class pins the byte-exact wire format documented at
// api/guides/authentication-advanced.md#test-vector and reproduced in
// tests/conformance/fixtures/auth-vector.json. The reference TypeScript
// implementation lives at packages/ts/src/core/internal/auth/bootstrap.ts;
// this file MUST reproduce the identical hex against the same inputs.
//
// Two-stage flow:
//   1. Serialize the request body as JSON, keys in source order
//      (keycode, email, deviceId), no whitespace, no trailing newline.
//   2. Build the canonical signing string and HMAC-SHA256 it with the
//      LicenseKey as the key.
//
// Why a dedicated module: the bootstrap signature predates any session
// (no sessionSecret exists yet), uses the LicenseKey as the HMAC key,
// and is the only call where DeviceId appears in the body. The
// steady-state session-signing helper (SignRequest) handles all other
// calls.
using System;
using System.Globalization;
using System.Security.Cryptography;
using System.Text;

namespace Isa.Sdk.Core;

/// <summary>
/// Inputs to the embedded HMAC bootstrap signature. Mirrors the
/// <c>auth-vector</c> fixture one-for-one. Property order matches the
/// canonical serialized-body order (<c>keycode, email, deviceId</c>);
/// the body is hand-serialized so neither field order nor JSON
/// whitespace can drift.
/// </summary>
public sealed record BootstrapInput(
    string Keycode,
    string Email,
    string LicenseKey,
    string DeviceId,
    string Method,
    string Path,
    long Timestamp);

/// <summary>
/// Output bundle. Returns every intermediate so conformance tests can
/// assert each stage independently — if a future regression flips the
/// serialized body, the failure points at exactly that stage instead
/// of just "hex differs".
/// </summary>
public sealed record BootstrapSignature(
    string SerializedBody,
    string Canonical,
    string Hex,
    string Header);

/// <summary>
/// Builds the embedded HMAC bootstrap signature for
/// <c>POST /v1/sessions</c>. Pinned bytewise by the
/// <c>tests/conformance/fixtures/auth-vector.json</c> fixture.
/// </summary>
public static class Bootstrap
{
    /// <summary>
    /// Builds the byte-exact bootstrap signature.
    /// </summary>
    /// <param name="input">Bootstrap inputs.</param>
    /// <returns>Every intermediate stage of the signing flow.</returns>
    /// <exception cref="ArgumentException">
    /// Thrown when any required field is blank or the timestamp is not positive.
    /// </exception>
    public static BootstrapSignature Build(BootstrapInput input)
    {
        if (input is null)
        {
            throw new ArgumentNullException(nameof(input));
        }
        Require(input.Keycode, nameof(input.Keycode));
        Require(input.Email, nameof(input.Email));
        Require(input.LicenseKey, nameof(input.LicenseKey));
        Require(input.DeviceId, nameof(input.DeviceId));
        Require(input.Method, nameof(input.Method));
        Require(input.Path, nameof(input.Path));
        if (input.Timestamp <= 0)
        {
            throw new ArgumentException(
                "bootstrap signature: timestamp must be positive",
                nameof(input.Timestamp));
        }

        var serializedBody = SerializeBody(input.Keycode, input.Email, input.DeviceId);
        var canonical = string.Concat(
            input.Timestamp.ToString(CultureInfo.InvariantCulture),
            ".",
            input.Method.ToUpperInvariant(),
            " ",
            input.Path,
            ".",
            serializedBody);

        var hex = ComputeHmacHex(input.LicenseKey, canonical);
        var header = "ISA-Signature: t="
            + input.Timestamp.ToString(CultureInfo.InvariantCulture)
            + ",v1=" + hex;

        return new BootstrapSignature(serializedBody, canonical, hex, header);
    }

    /// <summary>
    /// Hand-rolled JSON serialization to guarantee key order and no
    /// whitespace. System.Text.Json with PropertyNamingPolicy preserved
    /// would also work, but a hand-rolled writer is unambiguous: any
    /// future field addition is a deliberate edit here, not an
    /// accidental reorder elsewhere.
    /// </summary>
    private static string SerializeBody(string keycode, string email, string deviceId)
    {
        var sb = new StringBuilder(64 + keycode.Length + email.Length + deviceId.Length);
        sb.Append("{\"keycode\":");
        AppendJsonString(sb, keycode);
        sb.Append(",\"email\":");
        AppendJsonString(sb, email);
        sb.Append(",\"deviceId\":");
        AppendJsonString(sb, deviceId);
        sb.Append('}');
        return sb.ToString();
    }

    private static string ComputeHmacHex(string key, string canonical)
    {
        var keyBytes = Encoding.UTF8.GetBytes(key);
        var canonicalBytes = Encoding.UTF8.GetBytes(canonical);
        using var hmac = new HMACSHA256(keyBytes);
        var digest = hmac.ComputeHash(canonicalBytes);
        var sb = new StringBuilder(digest.Length * 2);
        for (var i = 0; i < digest.Length; i++)
        {
            sb.Append(digest[i].ToString("x2", CultureInfo.InvariantCulture));
        }
        return sb.ToString();
    }

    private static void AppendJsonString(StringBuilder sb, string s)
    {
        sb.Append('"');
        for (var i = 0; i < s.Length; i++)
        {
            var c = s[i];
            switch (c)
            {
                case '"':
                    sb.Append("\\\"");
                    break;
                case '\\':
                    sb.Append("\\\\");
                    break;
                case '\n':
                    sb.Append("\\n");
                    break;
                case '\r':
                    sb.Append("\\r");
                    break;
                case '\t':
                    sb.Append("\\t");
                    break;
                case '\b':
                    sb.Append("\\b");
                    break;
                case '\f':
                    sb.Append("\\f");
                    break;
                default:
                    if (c < 0x20)
                    {
                        sb.Append("\\u");
                        sb.Append(((int)c).ToString("x4", CultureInfo.InvariantCulture));
                    }
                    else
                    {
                        sb.Append(c);
                    }
                    break;
            }
        }
        sb.Append('"');
    }

    private static void Require(string value, string name)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            throw new ArgumentException(
                $"bootstrap signature: {name} is required",
                name);
        }
    }
}
